### PR TITLE
Default champion

### PR DESCRIPTION
**Pull request by Myrilandel**
## New Features
**DefaultChampion**
- This Champion is a Default one. It is loaded in game if we do not success to load a real champion.
- This is used is the dev' context.

**DefaultPassive**
- This Passiveis a Default one. It is loaded in game if we do not success to load a real Passive.
- This is used is the dev' context.

**DefaultAutoAttack**
- This AutoAttack is a Default one. It is loaded in game if we do not success to load a real AutoAttack .
- This is used is the dev' context.

**DefaultSpell**
- This Spell is a Default one. It is loaded in game if we do not success to load a real Spell.
- This is used is the dev' context.

**DefaultStatus**
- This Status is a Default one. It is loaded in game if we do not success to load a real Status.
- This is used is the dev' context.

## Enhancement
**Champion**
- Now able to check if we do not success to locate a Champion in the ChampionData.json
- If the loading fails, we load a DefaultChampion
- Now able to detect if one of the element that constitute the Champion (Spell,AA,Pasive) is broken. We replace Broken elements by a default one.

**PassiveBase**
- Now able to check if we do not success to load the data of a Passive in PassiveData.json
- If the loading fails, the Passive notifies a Champion to be replaced by a DefaultPassive.
- Now able to check if a Status, applied by the Passive itself, can not be loaded from the StatusData.json.
- If the loading of the Status fails, we replace it by a DefaultStatus.

**AutoAttackBase**
- Now able to check if we do not success to load the data of a AutoAttackin AutoAttackData.json
- If the loading fails, the AutoAttack notifies a Champion to be replaced by a DefaultAutoAttack.
- Now able to check if a Status, applied by the AutoAttack itself, can not be loaded from the StatusData.json.
- If the loading of the Status fails, we replace it by a DefaultStatus.

**Spell**
- Now able to check if we do not success to load the data of a Spellin SpellData.json
- If the loading fails, the Spellnotifies a Champion to be replaced by a DefaultSpell.
- Now able to check if a Status, applied by the Spellitself, can not be loaded from the StatusData.json.
- If the loading of the Status fails, we replace it by a DefaultStatus.

**IDisplayable**
- Members must have a bool IsLoaded property.

## KB Tasks & PlantUML
- [KB Task #85](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=85&project_id=3)

### CODE REVIEW

File: Assets/Scripts/Champions/BASE_CHAMPION/AutoAttackBase.cs [hunk 1]
```diff
@@ -23,45 +23,69 @@ public abstract class AutoAttackBase : MonoBehaviour, IDisplayable
...
+                    if (Status[i] == null || !Status[i].GetComponent<StatusBase>().PreWarm())
+                    {
+                        Debug.Log(AutoAttackDefinition.Status[i] + " can not be loaded. "
+                                 + "Please Ensure that the Status Name is correct in the SpellData.json file "
+                                 + "or that this Status exists as a Prefab with the same Script Name associated to it. "
+                                 + "DefaultStatus substitued");
+                        Status[i] = (GameObject)Resources.Load("Default/DefaultStatus");
+                        Status[i].GetComponent<StatusBase>().PreWarm();
+                    }
+                }
```
Comment (Veltouille) [inline Assets/Scripts/Champions/BASE_CHAMPION/AutoAttackBase.cs:71]: Celui qui a eu l'idée du IsLoaded est un génie ! Bien vu pour le Status si il se load mal alors qu'il a bien load l'attaque ;)

File: Assets/Scripts/Champions/BASE_CHAMPION/AutoAttackBase.cs [hunk 2]
```diff
@@ -23,45 +23,69 @@ public abstract class AutoAttackBase : MonoBehaviour, IDisplayable
...
+	 * Before everything, we check if the AutoAttack was correctly loaded from the JSON file. If it is not the case, we notify the Champion class to replace the broken AutoAttack by a DefaultAutoAttack.
+	 * Then, we initialize the CD of the spell.
 	 **/
     protected virtual void Start()
     {
+        if (!_isLoaded)
+        {
+            champion.ReplaceByDefaultDisplayable(this);
+            Debug.Log("Error when loading the json data of : " + this.GetType().Name + ". Please, check your AutoAttackData.Json. DefaultAutoAttack was substitued.");
+        }
```
Comment (Veltouille) [inline Assets/Scripts/Champions/BASE_CHAMPION/AutoAttackBase.cs:87]: Parfait !

File: Assets/Scripts/Champions/BASE_CHAMPION/Champion.cs [hunk 3]
```diff
@@ -175,44 +186,88 @@ protected virtual void AttributeAutoAttackToClass()
...
+
+        if (brokenDisplayable is AutoAttackBase)
+        {
+            autoAttack = (AutoAttackBase)gameObject.AddComponent(Type.GetType("DefaultAutoAttack"));
+            GUIAutoAttackDisplayer autoAttackDisplayer = GameObject.Find("AutoAttack").GetComponent<GUIAutoAttackDisplayer>();
+            autoAttackDisplayer.AttributeDisplayable(autoAttack);
+        }
+
+        Destroy((Behaviour)brokenDisplayable);
     }
```
Comment (Veltouille) [inline Assets/Scripts/Champions/BASE_CHAMPION/Champion.cs:271]: Super cette méthode hyper efficace grâce à l'archi ! Bien ouej ;)

File: Assets/Scripts/Champions/BASE_CHAMPION/PassiveBase.cs [hunk 4]
```diff
@@ -77,6 +108,7 @@ protected void LoadSpellData(string json)
                 if (passive.ScriptName == this.GetType().ToString())
                 {
                     PassiveDefinition = passive;
+                    _isLoaded = true;
                     break;
                 }
             }
```
Comment (Veltouille) [inline Assets/Scripts/Champions/BASE_CHAMPION/PassiveBase.cs:114]: La même que autoAttack

File: Assets/Scripts/Status/BASE_STATUS/StatusBase.cs [hunk 5]
```diff
@@ -47,25 +50,30 @@ protected virtual void Start()
...
+            TicksIntervals = StatusDefinition.TicksIntervals;
+            TickStarts = StatusDefinition.TickStarts;
+            Damages = StatusDefinition.Damages;
+            DamagesType = StatusDefinition.DamagesType;
+            OtherValues = StatusDefinition.OtherValues;
+            IsStackable = StatusDefinition.IsStackable;
+            NumberOfStacks = StatusDefinition.NumberOfStacks;
+            Description = StatusDefinition.Description;
+        }
+        return _isLoaded;
```
Comment (Veltouille) [inline Assets/Scripts/Status/BASE_STATUS/StatusBase.cs:76]: Joli ! Rien à dire

**Review by Veltouille — APPROVED** [review]
Et voilà qui est un excellent travail ! Dû a une archi de feu :D

Super boulot Jo ! Et un peu moi lol !
Je merge puisse qu'on avait tout vu ensemble !

File: Assets/Scripts/Champions/BASE_CHAMPION/Champion.cs (hunk 3 above)
Comment (Myrilandel) [reply Assets/Scripts/Champions/BASE_CHAMPION/Champion.cs:271]: Merci :D

**Comment by Myrilandel** [comment]
Merci pour tes commentaires et ton merge, je voulais juste rajouter que je suis parti de cette branche pour faire la GameJam donc pas de conflicts pour la suite normalement ;)

**Pull request merged**